libs/custom-scanner/src/custom_a4_scanner.ts

```diff
@@ -0,0 +1,632 @@
+        if (a4Status.imageHeightSideB !== 0) {
+          scannerImageB.imageHeight = a4Status.imageHeightSideB;
+        }
+        return ok('break');
+      }
+
+      if (
+        (scanSide === ScanSide.A || scanSide === ScanSide.A_AND_B) &&
+        !a4Status.endScanSideA
+      ) {
+        const processImageDataResult = await processImageData(ScanSide.A, {
+          pageSize: a4Status.pageSizeSideA,
+          imageWidth: a4Status.imageWidthSideA,
+          imageHeight: a4Status.imageHeightSideA,
+        });
+
+        /* istanbul ignore next */
+        if (processImageDataResult.isErr()) {
+          return processImageDataResult;
+        }
+      }
+
+      if (
+        (scanSide === ScanSide.B || scanSide === ScanSide.A_AND_B) &&
+        !a4Status.endScanSideB
+      ) {
+        const processImageDataResult = await processImageData(ScanSide.B, {
+          pageSize: a4Status.pageSizeSideB,
+          imageWidth: a4Status.imageWidthSideB,
+          imageHeight: a4Status.imageHeightSideB,
+        });
+
+        /* istanbul ignore next */
+        if (processImageDataResult.isErr()) {
+          return processImageDataResult;
+        }
+      }
+
+      /* istanbul ignore next */
+      if (
+        status.isScanInProgress &&
+        a4Status.pageSizeSideA === 0 &&
+        a4Status.pageSizeSideB === 0
+      ) {
+        debug('scan in progress but no data available');
+      }
+
+      return ok('continue');
+    };
+
+    try {
+      timer.checkpoint('start scan');
+      const startScanResult = await this.startScanInternal();
+
+      if (startScanResult.isErr()) {
+        return startScanResult;
+      }
+
+      for (;;) {
+        const result = await scanLoopTick();
+
+        if (result.isErr()) {
+          return result;
+        }
+
+        const action = result.ok();
+
+        /* istanbul ignore else */
+        if (action === 'continue') {
+          continue;
+        } else if (action === 'break') {
+          break;
+        } else {
+          throwIllegalValue(action);
+        }
+      }
+    } finally {
+      unlock();
+    }
+
+    timer.checkpoint('stop scan');
+    void (await this.stopScanInternal());
+
+    timer.end();
+    return ok([scannerImageA, scannerImageB]);
+  }
+
+  /**
+   * Gets part of the scanned image from the scanner. Does not hold the public
+   * lock.
+   */
+  private async getImagePortionBySideInternal(
+    scanSide: ScanSide,
+    imagePortionSize: number,
+    imageWidth: number,
+    imageHeight: number
+  ): Promise<Result<Buffer, ErrorCode>> {
+    const MAX_IMAGE_PORTION_SIZE = 0xffffff; // Max 3 bytes
+
+    /* istanbul ignore next */
+    if (imagePortionSize === 0) {
+      return ok(Buffer.alloc(0));
+    }
+
+    /* istanbul ignore next */
+    if (imagePortionSize > MAX_IMAGE_PORTION_SIZE) {
+      return await this.getImagePortionBySideInternal(
+        scanSide,
+        MAX_IMAGE_PORTION_SIZE,
+        imageWidth,
+        imageHeight
+      );
+    }
```

Comment:
Why do we cap the size here rather than erroring? Does this case actually occur?

Reply:
It's not an error when this happens. `imagePortionSize` here comes from `pageSizeSideA` which comes from `validPageSizeA`, which is a `u32` (4 bytes). When we request data from the scanner using the `GetImageDataRequest` code, the value goes into the `length` property, which is `u24` (3 bytes). So `MAX_IMAGE_PORTION_SIZE` is the most we can accept in a single request, but we're making a whole bunch of requests for image data anyway so we just keep asking for data until `validPageSizeA` is 0.

Reply:
Cool, maybe there's some way to make all of this more clear with some variable renaming or a comment. Not a big deal though. Thanks for helping me understand

Reply:
Okay, I added docs to all the fields in `StatusMessageInternal` and to `MAX_IMAGE_PORTION_SIZE`.